taichi/analysis/gather_uniquely_accessed_pointers.cpp

```diff
@@ -0,0 +1,157 @@
+  // top-level loop.
+  // If loop_unique_[stmt] is x > 0, the value of stmt is unique to
+  // the x-th loop index.
+  std::unordered_map<Stmt *, int> loop_unique_;
+
+ public:
+  int num_different_loop_indices{-1};
+  using BasicStmtVisitor::visit;
+
+  LoopUniqueStmtSearcher() {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(LoopIndexStmt *stmt) override {
+    if (stmt->loop->is<OffloadedStmt>())
+      loop_unique_[stmt] = stmt->index;
+  }
+
+  void visit(LoopUniqueStmt *stmt) override {
+    loop_unique_[stmt] = -1;
+  }
+
+  void visit(ConstStmt *stmt) override {
+    loop_invariant_.insert(stmt);
+  }
+
+  void visit(UnaryOpStmt *stmt) override {
+    if (loop_invariant_.count(stmt->operand) > 0) {
+      loop_invariant_.insert(stmt);
+    }
+    if (loop_unique_.count(stmt->operand) > 0 &&
+        (stmt->op_type == UnaryOpType::neg)) {
+      // TODO: Other injective unary operations
+      loop_unique_[stmt] = loop_unique_[stmt->operand];
+    }
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    if (loop_invariant_.count(stmt->lhs) > 0 &&
+        loop_invariant_.count(stmt->rhs) > 0) {
+      loop_invariant_.insert(stmt);
+    }
+    if ((loop_unique_.count(stmt->lhs) > 0 &&
+         loop_invariant_.count(stmt->rhs) > 0) &&
+        (stmt->op_type == BinaryOpType::add ||
+         stmt->op_type == BinaryOpType::sub ||
+         stmt->op_type == BinaryOpType::bit_xor)) {
+      loop_unique_[stmt] = loop_unique_[stmt->lhs];
+    }
+    if ((loop_invariant_.count(stmt->lhs) > 0 &&
+         loop_unique_.count(stmt->rhs) > 0) &&
+        (stmt->op_type == BinaryOpType::add ||
+         stmt->op_type == BinaryOpType::sub ||
+         stmt->op_type == BinaryOpType::bit_xor)) {
+      loop_unique_[stmt] = loop_unique_[stmt->rhs];
+    }
+  }
+
+  bool is_loop_unique(GlobalPtrStmt *stmt) const {
+    TI_ASSERT(num_different_loop_indices != -1);
+    std::vector<int> loop_indices;
+    loop_indices.reserve(stmt->indices.size());
+    for (auto &index : stmt->indices) {
+      auto loop_unique_index = loop_unique_.find(index);
+      if (loop_unique_index != loop_unique_.end()) {
+        if (loop_unique_index->second == -1) {
+          // LoopUniqueStmt
+          return true;
+        } else {
+          // LoopIndexStmt
+          loop_indices.push_back(loop_unique_index->second);
+        }
+      }
+    }
+    std::sort(loop_indices.begin(), loop_indices.end());
+    auto current_num_different_loop_indices =
+        std::unique(loop_indices.begin(), loop_indices.end()) -
+        loop_indices.begin();
+    // for i, j in x:
+    //     y[j, i] is loop-unique
+    //     a[i, i] is not loop-unique
+    return current_num_different_loop_indices == num_different_loop_indices;
+  }
+};
+
+class UniquelyAccessedSNodeSearcher : public BasicStmtVisitor {
+ private:
+  LoopUniqueStmtSearcher loop_unique_stmt_searcher_;
+  std::unordered_map<SNode *, GlobalPtrStmt *> accessed_pointer_;
```

Comment:
I'm a little confused about what `accessed_pointer_` means here.

Also, what if a `SNode *` is accessed by multiple `GlobalPtrStmt *`?

Reply:
> Also, what if a SNode * is accessed by multiple GlobalPtrStmt *?

Looking at L123 (`! definitely_same_address()`), I think when if there are multiple different `GlobalPtrStmt`s, then this SNode access is deemed as not unique?

Reply:
I agree the name is confusing... What about `uniquely_accessed_pointer_`?